### DISCUSSION TITLE
Request Input In Middleware

### DISCUSSION THREAD

**Discussion by Lermatroid** [next-safe-action/next-safe-action#149] [Q&A]
Hello!

I wanted to know if there is a way request input inside of a middleware on any actions that use that middleware. I am writing a middleware that performs mutations on documents, and want to request a document ID every time and validate its existence / that the user owns it. 

Is this something I can implement in middleware? Thanks!

**Comment by TheEdoRan** [comment]
This would be the _easy_ way:

```typescript
import { z } from "zod";
import { actionClient } from "@/lib/safe-action";

const documentSchema = z.object({
  documentId: z.string(),
});

const testAction = actionClient
  .use(async ({ next, ctx, clientInput }) => {
    const parsedInput = clientInput as z.infer<typeof documentSchema>;

    // Access document id
    const { documentId } = parsedInput;

    return next({ ctx });
  })
  .schema(documentSchema)
  .action(async () => {});
```

You could export `documentSchema` and use it with all the actions that require a `documentId`, by extending/merging the object with Zod.
